### ISSUE TITLE
Add `hal::stream_dac`

### ISSUE THREAD

**Issue by kammce** [libhal/libhal#45]
This interface would allow a user to pass a span of dac values to a dac driver. The dac driver should use DMA to manage the stream to the peripheral. Interrupts can be used as well but ill advised for performance reasons. And polling can be done, but it would make the interface less useful in general.

**Comment by kammce** [comment]
# List of MCUs with DMA DACs

Each with a page linking to the dac data register (or nearby it).

- [msp430](https://www.ti.com/lit/ug/slau144k/slau144k.pdf?ts=1717333740042&ref_url=https%253A%252F%252Fwww.google.com%252F) page 647
    - Justification: Right ❌
    - Resolution: 8-bit + 12 -bit
- [stm32f1](https://github.com/libhal/libhal-stm32f1/blob/main/datasheets/RM0008.pdf) page 257
    - Justification: Left & Right
    - Resolution: 8-bit + 12 -bit
- [lpc40](https://github.com/libhal/libhal-lpc40/blob/main/datasheets/UM10562.pdf) page 815
    - Justification: Left
    - Resolution: 10-bit
- [samd21](https://cdn.sparkfun.com/datasheets/Dev/Arduino/Boards/Atmel-42181-SAM-D21_Datasheet.pdf) page 928
    - Justification: Left & Right
    - Resolution: 10-bit
- [samd20](https://ww1.microchip.com/downloads/en/DeviceDoc/SAM_D20_%20Family_Datasheet_DS60001504C.pdf) page 548
    - Justification: Left & Right
    - Resolution: 10-bit
- [esp32](https://www.espressif.com/sites/default/files/documentation/esp32_technical_reference_manual_en.pdf#sensor) page 641
    - Justification: N/A (aligned)
    - Resolution: 8-bit
- [pic24f](https://ww1.microchip.com/downloads/en/DeviceDoc/39615A.pdf) page 6
    - Justification: Left & Right
    - Resolution: 10-bit

## Devices considered without DAC

These are devices that were looked into but did not have a DAC. This is here to document which devices were looked at so others will not attempt it later.

- RP2040
- stm32f411

**Comment by kammce** [comment]
# Proposed interface

```C++
#pragma once

#include <concepts>
#include <span>

#include "units.hpp"

namespace hal {
/**
 * @brief Hardware abstraction for a dac that accepts a stream of data
 *
 * This interface can be used for the following:
 *
 * 1. Audio output using PCM8 or PCM16 streams
 * 2. Arbitrary waveform generation
 * 3. Single wire data protocols (ex: neopixel LEDs)
 *
 * The driver is likely to use the following to operate:
 *
 * 1. DMA: preferred choice as it removes the need for the CPU to handle the
 *    transfer.
 * 2. SPI (using DMA): preferred when communicating with an external dac
 * 3. Interrupts & Timers: Not recommended but can be used. The CONs of this
 *    approach is a low maximum sample rate, sample rate accuracy due to
 *    interrupt & software latency, and requires cycles from the CPU.
 *
 * @tparam data_t - container size for the sample data. For such things as PCM8
 * and PCM16 this would be std::uint8_t and std::uint16_t respectfully.
 */
template<std::unsigned_integral data_t>
class stream_dac
{
public:
  struct samples
  {
    /**
     * Sample rate sets the update rate of the dac. For example, if the sample
     * rate is 44.1kHz (typical for consumer audio), then the samples will be
     * read from the data array and written to the dac output at that frequency.
     *
     */
    hal::hertz sample_rate;
    /**
     * Span of unsigned integer data to be written to the dac at the rate
     * specified by the sample rate field.
     *
     * If this is empty (empty() == true), then the write command simply
     * returns without updating anything.
     */
    std::span<data_t> data;
  };

  /**
   * @brief Write a sequence of data to the dac
   *
   * This API is blocking call. The expectation is that implementors
   *
   *
   * @param p_samples - sample data to be written to the dac output
   * @throws hal::argument_out_of_domain - when the sample rate is not possible
   * for the driver. This can happen if:
   *     - The sample rate is above the input clock rate of the dac making it
   *       impossible to produce.
   *     - The dac could not reach a sample rate close enough to the desired
   *       sample rate. This can occur if the dac's source clock and the sample
   *       rates are close and as such no divider cleanly reaches a reasonable
   *       sample rate for the device. For example, if the desired sample rate
   *       is 750kHz and the clock rate is 1MHz and divider is an integer. You
   *       would either operate at 1MHz or 500kHz which are very far away from
   *       the desired sample rate. The driver should, but is not required, to
   *       provide a means to control the allowed "error" between the desired
   *       sample rate and the achievable sample rate in order tune what
   *       samples are rejected and an exception thrown.
   *     - Sample rate is too low and cannot be achieved by the dac.
   */
  void write(samples p_samples)
  {
    driver_write(p_samples);
  }

private:
  virtual void driver_write(samples p_samples) = 0;
};
}  // namespace hal
```

**Comment by kammce** [comment]
Meant to make the span const. I considered leaving it mutable such that a driver can mutate the memory in place in order to arrange the data in such a way that a right justified DAC register could work with DMA. 

But this would be a bad API design. It means that any buffer you pass into these APIs is liable to destroy/truncate the data. Which would require the user to make duplicate buffers, one for each copy of the buffer it wants to transfer over DAC or to other stream drivers. Using const allows us to pass the same buffer to every stream API without concern. 

Also this would impact the ability to stream data from FLASH and rodata sections without a copy. 

Prevent the data from being modified is the right choice here.

**Comment by kammce** [comment]
When I used audacity I realized that PCM 8-bit unsigned is a thing, but PCM 16 bits and above is signed. This is a bit of an issue since all of the DACs are unsigned. Will have to consider what is reasonable for this.

**Comment by kammce** [comment]
I think its fine to require unsigned data to be passed into the DAC. For example, if an application is converting MP3 data into signed 16-bit PCM, then the code will also have to perform a step to convert those samples to unsigned PCM16 by adding (2^15 - 1) to each of the values in the set. Then pass that resulting info to the dac.

**Comment by kammce** [comment]
Resolved by #48
